project/BuildDependencies/DownloadMingwBuildEnv.bat

```diff
@@ -1,63 +1,568 @@
-IF EXIST %TMP_PATH% rmdir %TMP_PATH% /S /Q
+for %%b in (%1, %2, %3) do (
+  if %%b==build64 (set build64=yes)
+  if %%b==sh (set opt=sh)
+)
 
-IF $%1$ == $$ (
-  SET DL_PATH="%BUILD_DEPS_PATH%\downloads2"
-) ELSE (
-  SET DL_PATH="%1"
+:: use 32bit msys2 on x86 machine
+if %PROCESSOR_ARCHITECTURE%=="x86" set msys2=msys32
+if %msys2%==msys32 (set arch=i686) else (set arch=x86_64)
+set msysfile=msys2-base-%arch%-%msysver%.tar.xz
+if %opt%==mintty (
+    set sh=%instdir%\%msys2%\usr\bin\mintty.exe -d -i /msys2.ico /usr/bin/bash
+) else (
+    set sh=%instdir%\%msys2%\usr\bin\sh.exe
 )
 
-SET WGET=%BUILD_DEPS_PATH%\bin\wget
-SET ZIP=%BUILD_DEPS_PATH%\..\Win32BuildSetup\tools\7z\7za
+::------------------------------------------------------------------
+::download and install basic msys2 system:
+::------------------------------------------------------------------
+if exist "%instdir%\%msys2%\msys2_shell.bat" GOTO minttySettings
+	if not exist %downloaddir% mkdir %downloaddir%
+
+:download
+if exist "%downloaddir%\%msysfile%" (
+    setlocal EnableDelayedExpansion
+    for /F "tokens=*" %%A in ("%downloaddir%\%msysfile%") do set fileSize=%%~zA
+    if !fileSize!==0 del %downloaddir%\%msysfile%
+    endlocal
+    )
+
+if exist "%downloaddir%\%msysfile%" GOTO unpack
+    echo -------------------------------------------------------------------------------
+    echo.- Download msys2 basic system (Kodi mirrors: %usemirror%)
+    echo -------------------------------------------------------------------------------
+
+    set msysurl=http://sourceforge.net/projects/msys2/files/Base/%arch%/%msysfile%/download
+    if %usemirror%==yes (
+        ::download msys2 from our mirror
+        set msysurl=%KODI_MIRROR%/%msysfile%
+    )
+    %instdir%\bin\wget --tries=20 --retry-connrefused --waitretry=2 --no-check-certificate -c -O %downloaddir%\%msysfile% %msysurl%
+    if errorlevel == 1 (
+        if exist "%downloaddir%\%msysfile%" del %downloaddir%\%msysfile%
+        if %usemirror%==yes (
+            set usemirror=no
+            goto download
+        )
+        echo ERROR: Unable to download msys2!
+        exit /B 1
+    )
+
+:unpack
+if exist "%downloaddir%\%msysfile%" (
+    echo -------------------------------------------------------------------------------
+    echo.- Install msys2 basic system
+    echo -------------------------------------------------------------------------------
+	%unpack_exe% x %downloaddir%\%msysfile% -so | %unpack_exe% x -aoa -si -ttar -o%instdir%
+	)
+	
+if not exist %instdir%\%msys2%\usr\bin\msys-2.0.dll (
+	echo -------------------------------------------------------------------------------
+	echo.- Install msys2 basic system failed,
+	echo -------------------------------------------------------------------------------
+	exit /B 1
+	)
+
+:minttySettings
+if exist %instdir%\%msys2%\home\%USERNAME%\.minttyrc GOTO updatemirrors
+if not exist %instdir%\%msys2%\home\%USERNAME% mkdir %instdir%\%msys2%\home\%USERNAME%
```

Comment:
disagree. Actually the main idea to generate all "configs" is do not support a lot of files. If something was changed in the bat file then the generated config will be changed also. I don't like idea support a lot of small config files instead of one bat file. look at the history of this bat file https://github.com/xbmc/xbmc/commits/master/project/BuildDependencies/DownloadMingwBuildEnv.bat
it was updated about once in a year. same will be for the new version.